llama_index/llms/openai_utils.py

```diff
@@ -326,12 +327,44 @@ def resolve_openai_credentials(
     api_base = api_base or openai.api_base or DEFAULT_OPENAI_API_BASE
     api_version = api_version or openai.api_version or DEFAULT_OPENAI_API_VERSION
 
-    if not api_key:
+    if not api_key and api_type not in ("azuread", "azure_ad"):
         raise ValueError(MISSING_API_KEY_ERROR_MESSAGE)
 
     return api_key, api_type, api_base, api_version
 
 
+def refresh_openai_azuread_token(
+    azure_ad_token: Any = None,
+) -> Any:
+    """
+    Checks the validity of the associated token, if any, and tries to refresh it
+    using the credentials available in the current context. Different authentication
+    methods are tried, in order, until a successful one is found as defined at the
+    package `azure-indentity`.
+    """
+    try:
+        from azure.core.exceptions import ClientAuthenticationError
+        from azure.identity import DefaultAzureCredential
+    except ImportError as ex:
+        raise ValueError(
+            "Using API type `azure_ad` or `azuread` requires the package"
+            " `azure-identity` to be installed."
+        ) from ex
+
+    if not azure_ad_token or azure_ad_token.expires_on < time.time() + 60:
+        try:
+            credential = DefaultAzureCredential()
+            azure_ad_token = credential.get_token(
+                "https://cognitiveservices.azure.com/.default"
```

Comment:
@santiagxf : I have a custom (private) scope. Shouldn't this be an argument of the function to allow users to configure it accordingly?

Reply:
@younes-io, Azure OpenAI falls under Cognitive Services scope. How are you using a custom scope with Azure OpenAI?

Reply:
@santiagxf : I was provided with a scope value like this: `api://xxxxxf-3aaa-xxxxx-xxxxx-81xxxxxxxxxxc/.default`

Reply:
I couldn't find any documentation pointing towards that type of scopes. It may be related with a private networking (looks like a private endpoint) but I couldn't find any doc confirming that such configuration is required. Do you have any pointers?